# Initial Data Exploration

This notebook does some early cleaning of a data frame with indicators for child well-being in OECD member nations and those same indicators for some nations who are not in the OECD. This notebook also gives a general look at the distribution of some of the variables. The data was pulled for the [OECD's database](https://stats.oecd.org/) underneath the section on "Social Protection and Well-Being" which also includes data on poverty/wealth distribution, gender, and general well-being. I wanted any data I used to be recent and I wanted to do some temporal analysis, so I chose data from 2014 to 2018 to explore. These explorations include: 
* How many observations and variables are there?
* How many data values are there for each year in the 5-year period?
* How many data values are there for each indicator of child-well-being?
* For each country in the data set, which year provides the most data values? 
* For each indicator, how many values are there per year?

---------

#### HISTORY

* 10/25/20 aheyward - Initial exploration of OECD dataset

In [1]:
#Importing necessary packages
import pandas as pd

In [2]:
#Importing the raw data from the data folder
oecd_df = pd.read_csv("../data/raw/OECD_child_wellbeing.csv")

#### How many observations and variables are there?
By pulling the shape of the data, we can see how many observation it has in its raw state.

In [3]:
#Getting the shape of the DataFrame rem_df and use two pointers 
ncol = oecd_df.shape[1] #indexing the number of columns
nrow = oecd_df.shape[0] #indexing the number of rows

print(ncol)
print(nrow)

9
2335


There are 2,153 observations in this data set, but there are only 37 OECD member countries, in addition to data from 10 non-OECD nations. Where do all these extra observations come from? I will run `head()` to gauge the unit of analysis. 

In [4]:
oecd_df.head(10)

COU    Country     IND                                          Indicator  \
0  AUS  Australia    CWB3  Average disposable household income, 0-17 year...   
1  AUS  Australia    CWB3  Average disposable household income, 0-17 year...   
2  AUS  Australia   CWB8A  Adolescents (15-year-olds) who report talking ...   
3  AUS  Australia   CWB16  Adolescents (15-year-olds) skipping either bre...   
4  AUS  Australia  CWB17A  Adolescents (15-year-olds) who report doing no...   
5  AUS  Australia  CWB17B  Adolescents (15-year-olds) who report regularl...   
6  AUS  Australia   CWB22  Children (3-5 year-olds) enrolled in pre-prima...   
7  AUS  Australia   CWB22  Children (3-5 year-olds) enrolled in pre-prima...   
8  AUS  Australia   CWB22  Children (3-5 year-olds) enrolled in pre-prima...   
9  AUS  Australia   CWB22  Children (3-5 year-olds) enrolled in pre-prima...   

   YEAR  Year    Value  Flag Codes  Flags  
0  2014  2014  34872.4         NaN    NaN  
1  2016  2016  34788.7         NaN    NaN  
2  2015  2015     96.7         NaN    NaN  
3  2015  2015     22.1         NaN    NaN  
4  2015  2015      6.5         NaN    NaN  
5  2015  2015     52.5         NaN    NaN  
6  2014  2014     85.3         NaN    NaN  
7  2015  2015     86.5         NaN    NaN  
8  2016  2016     84.7         NaN    NaN  
9  2017  2017     84.0         NaN    NaN

In [5]:
oecd_df.columns

Index(['COU', 'Country', 'IND', 'Indicator', 'YEAR', 'Year', 'Value',
       'Flag Codes', 'Flags'],
      dtype='object')

#### How many values are there in each column?
Here you can see that each row is the data value for one indicator of child well-being for a specific country in a given year. How many countries, years, and indicators are included int the data?

In [6]:
oecd_df.nunique()

COU            50
Country        50
IND            29
Indicator      29
YEAR            5
Year            5
Value         975
Flag Codes      0
Flags           0
dtype: int64

There appear to be 50 countries in the data, data from 5 different years, and 29 indicators. Despite there being 2,335 observations, there are only 975 actual data values, which means there must be a great deal of missing data. There are also two variables for which no data is recorded at all. Which year had the data recorded for the national indicators?

In [7]:
oecd_df['Year'].value_counts()

2015    1048
2014     395
2016     341
2017     313
2018     238
Name: Year, dtype: int64

2015 has the most data out of the 5 years. Breaking it down further below by country, you can see that not all countries have data for a certain indicator, and when they do, it is not consistent across each year. For example, below you can see that Australia has values for 4 indicators in 2014, while the US 9 for the same year. However, the US also has a whopping 24 values recorded for the following year, 2015. 

In [8]:
oecd_df.groupby('Country')['Year'].value_counts()

Country        Year
Australia      2015    17
               2014     4
               2016     4
               2017     3
               2018     3
                       ..
United States  2015    24
               2014     9
               2016     9
               2017     9
               2018     8
Name: Year, Length: 235, dtype: int64

### Cleaning the data frame
Here, we renamed the columns so they make more sense and have a uniform format.
Then, we subsetted the original data frame to get rid of columns that were nothing but missing values for each observation and applied the new column names.

In [9]:
#Creating a list that has only the columns going into the new data frame.
cols_to_use = ['COU', 'Country', 'IND', 'Indicator', 'YEAR', 'Value']

#Creating a dictionary with new column names (values) that correspond with the original column names (keys)
col_map = {'COU': 'country_code',
           'Country': 'country',
           'IND': 'indicator_code',
           'Indicator': 'indicator',
           'YEAR': 'year',
           'Value': 'data_value'}

#Filtering the original data frame by the list of columns and applying the dictionary values
oecd_df2 = oecd_df[cols_to_use].rename(columns=col_map)

In [10]:
#Checkin the new layout of the data frame
oecd_df2.head(5)

country_code    country indicator_code  \
0          AUS  Australia           CWB3   
1          AUS  Australia           CWB3   
2          AUS  Australia          CWB8A   
3          AUS  Australia          CWB16   
4          AUS  Australia         CWB17A   

                                           indicator  year  data_value  
0  Average disposable household income, 0-17 year...  2014     34872.4  
1  Average disposable household income, 0-17 year...  2016     34788.7  
2  Adolescents (15-year-olds) who report talking ...  2015        96.7  
3  Adolescents (15-year-olds) skipping either bre...  2015        22.1  
4  Adolescents (15-year-olds) who report doing no...  2015         6.5

In [11]:
#Saving the data frame as a .csv file in the `cleaned` data folder
oecd_df2.to_csv('../data/cleaned/CLEAN_OECD_child_wellbeing.csv', index=False)

Next, it would be good to get a sense of how many indicators are present for each country in a given year.

In [12]:
#A grouped chart to see differences in data across time.
indicators_by_loc_and_year = oecd_df2.groupby(['country_code', 'year'])['indicator_code'].count().unstack()
indicators_by_loc_and_year

year          2014  2015  2016  2017  2018
country_code                              
AUS            4.0  17.0   4.0   3.0   3.0
AUT           12.0  26.0  10.0   9.0   8.0
BEL            9.0  26.0  10.0   9.0   8.0
BEL-VLG        NaN   1.0   NaN   NaN   NaN
BGR            9.0  23.0   7.0   7.0   6.0
BRA            1.0  17.0   1.0   1.0   1.0
CAN            5.0  16.0   6.0   3.0   2.0
CHE           10.0  23.0   7.0   6.0   3.0
CHL            3.0  21.0   3.0   4.0   1.0
COL            NaN  17.0   1.0   1.0   NaN
CRI            2.0  18.0   4.0   4.0   1.0
CYP            7.0   7.0   7.0   7.0   5.0
CZE           12.0  26.0  10.0   9.0   8.0
DEU           11.0  25.0   9.0   8.0   7.0
DNK           12.0  24.0  10.0   9.0   8.0
ESP           12.0  27.0  10.0   9.0   8.0
EST           12.0  26.0  10.0   9.0   8.0
FIN           12.0  26.0  10.0  10.0   8.0
FRA           12.0  27.0  10.0   9.0   5.0
GBR           10.0  26.0   9.0  10.0   5.0
GRC           12.0  26.0  10.0   9.0   8.0
HRV            9.0  24.0   7.0   7.0   6.0
HUN           12.0  26.0  10.0   9.0   5.0
IDN            1.0   4.0   1.0   1.0   NaN
IRL           12.0  27.0  10.0   9.0   5.0
ISL            8.0  24.0   6.0   3.0   3.0
ISR            8.0  17.0   6.0   6.0   3.0
ITA           12.0  25.0  10.0   9.0   8.0
JPN            3.0  23.0   3.0   3.0   3.0
KOR            3.0  21.0   4.0   4.0   3.0
LTU           10.0  26.0  10.0   9.0   8.0
LUX           11.0  27.0  10.0   9.0   5.0
LVA           12.0  26.0  10.0   9.0   8.0
MEX            4.0  20.0   4.0   3.0   3.0
MLT            9.0  11.0   6.0   6.0   6.0
NLD           12.0  26.0  10.0   9.0   7.0
NOR           10.0  21.0   8.0   8.0   5.0
NZL            3.0  17.0   3.0   3.0   3.0
OAVG           3.0  18.0   1.0   1.0   1.0
PER            NaN  16.0   NaN   NaN   NaN
POL           11.0  26.0  10.0   9.0   8.0
PRT           12.0  27.0  10.0   9.0   7.0
ROU            9.0  11.0   7.0   7.0   5.0
RUS            2.0  17.0   2.0   1.0   1.0
SVK           12.0  25.0  10.0   9.0   5.0
SVN           12.0  26.0  10.0   9.0   8.0
SWE           12.0  24.0  10.0  10.0   8.0
TUR            7.0  22.0   6.0   6.0   3.0
USA            9.0  24.0   9.0   9.0   8.0
ZAF            NaN   2.0   NaN   NaN   NaN

In [13]:
#A bar graph to quickly visualize which year has the most data available.
indicators_by_loc_and_year.sort_index(ascending=False).plot(kind='barh',figsize=(6,20))

<AxesSubplot:ylabel='country_code'>

Given the results of the bar graph, it again appears that 2015 has the most data available by far. There is very little data available for any year for non-OCED  nations.
* Indonesia does not have many indicators for any year, and Peru only has data for indicators in 2015. South Africa and Flanders also only have data for 2015, but there are less than 5 values recorded for either.
* For each country, all of the other years have fewer than 15 data values, suggesting that the other four years do not offer very much data for substantial analysis.

#### How many values are there per data indicator?

In [14]:
oecd_df2.groupby(['indicator_code', 'indicator']).size().sort_values(ascending=False).to_frame()

0
indicator_code indicator                                              
CWB50B         Length of paid paternity and parental leave res...  181
CWB50A         Length of paid maternity and parental leave ava...  181
CWB22          Children (3-5 year-olds) enrolled in pre-primar...  174
CWB1A          Children (0-17) living with two parents (%)         157
CWB1B          Children (0-17) living with a single parent (%)     157
CWB1C          Children (0-17) living in 'other' types of hous...  157
CWB2A          Children (0-14) in households where all adults ...  149
CWB2C          Children (0-14) in households with all adults n...  149
CWB3           Average disposable household income, 0-17 year-...  114
CWB9           Children (0-17) in overcrowded households (%)       113
CWB40          Adolescents (15-year-olds) who are high or extr...   77
CWB35B         Mathematics performance at age 15 (PISA)             47
CWB35A         Reading performance at age 15 (PISA)                 47
CWB33          Adolescents (15-year-olds) who feel like outsid...   46
CWB32          Adolescents (15-year-olds) who feel like they b...   46
CWB29          Adolescents (15-year-olds) who want top grades ...   44
CWB28          Adolescents (15-year-olds) who study before or ...   44
CWB16          Adolescents (15-year-olds) skipping either brea...   44
CWB30          Adolescents (15-year-olds) who feel anxious abo...   44
CWB8A          Adolescents (15-year-olds) who report talking t...   44
CWB27          Adolescents (15-year-olds) with parents interes...   43
CWB17B         Adolescents (15-year-olds) who report regularly...   43
CWB17A         Adolescents (15-year-olds) who report doing no ...   43
CWB31          Adolescents (15-year-olds) who expect to comple...   43
CWB44A         Adolescents (15-year-olds) reporting high life ...   37
CWB44B         Adolescents (15-year-olds) reporting low life s...   37
CWB51A         Children (11-, 13- and 15-year-olds) who report...   30
CWB51B         Children (11-, 13- and 15-year-olds) who report...   30
CWB8B          Adolescents (15-year-olds) whose parents report...   14

The indicators that look most promising because they have the most data are length of maternity leave, info about primary school, information about parental employment, and indicators about familial structure. All of these have well over 100 values recorded.

#### For each indicator, how many values are there per year?

In [15]:
oecd_df2.groupby(['indicator_code', 'indicator','year']).size().sort_values(ascending=False).unstack()

year                                                               2014  2015  \
indicator_code indicator                                                        
CWB16          Adolescents (15-year-olds) skipping either brea...   NaN  44.0   
CWB17A         Adolescents (15-year-olds) who report doing no ...   NaN  43.0   
CWB17B         Adolescents (15-year-olds) who report regularly...   NaN  43.0   
CWB1A          Children (0-17) living with two parents (%)         33.0  34.0   
CWB1B          Children (0-17) living with a single parent (%)     33.0  34.0   
CWB1C          Children (0-17) living in 'other' types of hous...  33.0  34.0   
CWB22          Children (3-5 year-olds) enrolled in pre-primar...  40.0  45.0   
CWB27          Adolescents (15-year-olds) with parents interes...   NaN  43.0   
CWB28          Adolescents (15-year-olds) who study before or ...   NaN  44.0   
CWB29          Adolescents (15-year-olds) who want top grades ...   NaN  44.0   
CWB2A          Children (0-14) in households where all adults ...  30.0  30.0   
CWB2C          Children (0-14) in households with all adults n...  30.0  30.0   
CWB3           Average disposable household income, 0-17 year-...  34.0  36.0   
CWB30          Adolescents (15-year-olds) who feel anxious abo...   NaN  44.0   
CWB31          Adolescents (15-year-olds) who expect to comple...   NaN  43.0   
CWB32          Adolescents (15-year-olds) who feel like they b...   NaN  46.0   
CWB33          Adolescents (15-year-olds) who feel like outsid...   NaN  46.0   
CWB35A         Reading performance at age 15 (PISA)                 NaN  47.0   
CWB35B         Mathematics performance at age 15 (PISA)             NaN  47.0   
CWB40          Adolescents (15-year-olds) who are high or extr...   NaN  39.0   
CWB44A         Adolescents (15-year-olds) reporting high life ...   NaN  37.0   
CWB44B         Adolescents (15-year-olds) reporting low life s...   NaN  37.0   
CWB50A         Length of paid maternity and parental leave ava...  36.0  36.0   
CWB50B         Length of paid paternity and parental leave res...  36.0  36.0   
CWB51A         Children (11-, 13- and 15-year-olds) who report...  30.0   NaN   
CWB51B         Children (11-, 13- and 15-year-olds) who report...  30.0   NaN   
CWB8A          Adolescents (15-year-olds) who report talking t...   NaN  44.0   
CWB8B          Adolescents (15-year-olds) whose parents report...   NaN  14.0   
CWB9           Children (0-17) in overcrowded households (%)       30.0  28.0   

year                                                               2016  2017  \
indicator_code indicator                                                        
CWB16          Adolescents (15-year-olds) skipping either brea...   NaN   NaN   
CWB17A         Adolescents (15-year-olds) who report doing no ...   NaN   NaN   
CWB17B         Adolescents (15-year-olds) who report regularly...   NaN   NaN   
CWB1A          Children (0-17) living with two parents (%)         34.0  32.0   
CWB1B          Children (0-17) living with a single parent (%)     34.0  32.0   
CWB1C          Children (0-17) living in 'other' types of hous...  34.0  32.0   
CWB22          Children (3-5 year-olds) enrolled in pre-primar...  44.0  45.0   
CWB27          Adolescents (15-year-olds) with parents interes...   NaN   NaN   
CWB28          Adolescents (15-year-olds) who study before or ...   NaN   NaN   
CWB29          Adolescents (15-year-olds) who want top grades ...   NaN   NaN   
CWB2A          Children (0-14) in households where all adults ...  30.0  30.0   
CWB2C          Children (0-14) in households with all adults n...  30.0  30.0   
CWB3           Average disposable household income, 0-17 year-...  33.0  11.0   
CWB30          Adolescents (15-year-olds) who feel anxious abo...   NaN   NaN   
CWB31          Adolescents (15-year-olds) who expect to comple...   NaN   NaN   
CWB32          Adolescents (15-year-olds) who feel like they b...   NaN   NaN   
CWB33          Adolescents (15

Examining this further by year, the reason why the aforementioned variables have so many values recorded is because they are the only ones where information was recorded across all years. 